cubeb-backend/src/log.rs

```diff
@@ -5,69 +5,108 @@
 
 #[macro_export]
 macro_rules! cubeb_log_internal {
-    ($level: expr, $msg: expr) => {
+    ($use_async: ident, $level: expr, $msg: expr) => {
         #[allow(unused_unsafe)]
         unsafe {
             if $level <= $crate::ffi::g_cubeb_log_level.into() {
-                cubeb_log_internal!(__INTERNAL__ $msg);
+                cubeb_log_internal!($use_async, __INTERNAL__ $msg);
             }
         }
     };
-    ($level: expr, $fmt: expr, $($arg: expr),+) => {
+    ($use_async: ident, $level: expr, $fmt: expr, $($arg: expr),+) => {
         #[allow(unused_unsafe)]
         unsafe {
             if $level <= $crate::ffi::g_cubeb_log_level.into() {
-                cubeb_log_internal!(__INTERNAL__ format!($fmt, $($arg),*));
+                cubeb_log_internal!($use_async, __INTERNAL__ format!($fmt, $($arg),*));
             }
         }
     };
-    (__INTERNAL__ $msg: expr) => {
-        if let Some(log_callback) = $crate::ffi::g_cubeb_log_callback {
+    ($use_async: ident, __INTERNAL__ $msg: expr) => {
+        fn get_cstr<'a>(buf: &'a mut [u8; 1024], file: &str, line: u32, msg: &str) -> &'a std::ffi::CStr {
             use std::io::Write;
-
-            let mut buf = [0 as u8; 1024];
-            let filename = std::path::Path::new(file!())
+            let filename = std::path::Path::new(file)
                 .file_name()
                 .unwrap()
                 .to_str()
                 .unwrap();
             // 2 for ':', 1 for ' ', 1 for '\n', and 1 for converting `line!()` to number of digits
-            let len = filename.len() + ((line!() as f32).log10().trunc() as usize) + $msg.len() + 5;
+            let len = filename.len() + ((line as f32).log10().trunc() as usize) + msg.len() + 5 + 5;
             debug_assert!(len < buf.len(), "log will be truncated");
-            let _ = write!(&mut buf[..], "{}:{}: {}\n", filename, line!(), $msg);
+            let _ = write!(&mut buf[..], "{}:{}: {}\n", filename, line, msg);
             let last = std::cmp::min(len, buf.len() - 1);
             buf[last] = 0;
             let cstr = unsafe { std::ffi::CStr::from_bytes_with_nul_unchecked(&buf[..=last]) };
-            log_callback(cstr.as_ptr());
+            cstr
+        }
+
+        match($use_async) {
```

Comment:
Since `$use_async` is an `ident`, could we use a literal for `$use_async` to distinguish whether the call is an async log or not (see https://doc.rust-lang.org/reference/macros-by-example.html#transcribing), which should have better readability? What do you think?